providers/mlx5/verbs.c

```diff
@@ -154,13 +154,102 @@ struct ibv_pd *mlx5_alloc_pd(struct ibv_context *context)
 	return &pd->ibv_pd;
 }
 
+static void mlx5_put_bfreg_index(struct mlx5_context *ctx, uint32_t bfreg_dyn_index)
+{
+	pthread_mutex_lock(&ctx->dyn_bfregs_mutex);
+	ctx->count_dyn_bfregs[bfreg_dyn_index]--;
+	pthread_mutex_unlock(&ctx->dyn_bfregs_mutex);
+}
+
+static int mlx5_get_bfreg_index(struct mlx5_context *ctx)
+{
+	int i;
+
+	pthread_mutex_lock(&ctx->dyn_bfregs_mutex);
+	for (i = 0; i < ctx->num_dyn_bfregs; i++) {
+		if (!ctx->count_dyn_bfregs[i]) {
+			ctx->count_dyn_bfregs[i]++;
+			pthread_mutex_unlock(&ctx->dyn_bfregs_mutex);
+			return i;
+		}
+	}
+
+	pthread_mutex_unlock(&ctx->dyn_bfregs_mutex);
+
+	return -1;
+}
+
+/* Returns a dedicated BF to be used by a thread domain */
 static struct mlx5_bf *mlx5_attach_dedicated_bf(struct ibv_context *context)
 {
+	struct mlx5_uar_info uar;
+	struct mlx5_context *ctx = to_mctx(context);
+	struct mlx5_device *dev = to_mdev(ctx->ibv_ctx.device);
+	int bfreg_dyn_index;
+	uint32_t bfreg_total_index;
+	uint32_t uar_page_index;
+	int index_in_uar, index_uar_in_page;
+	int mmap_bf_index;
+	int num_bfregs_per_page;
+
+	bfreg_dyn_index = mlx5_get_bfreg_index(ctx);
+	if (bfreg_dyn_index < 0) {
+		errno = ENOENT;
+		return NULL;
+	}
+
+	bfreg_total_index = ctx->start_dyn_bfregs_index + bfreg_dyn_index;
+	/* Check whether this bfreg index was already mapped and ready to be used */
+	if (ctx->bfs[bfreg_total_index].reg)
```

Comment:
@yishaih is this just a sanity check or is there a scenario where this would evaluate to true? `mlx5_get_bfreg_index()` returns an unused `bfreg` (whose count is 0). Locks on `dyn_bfregs_mutex` ensures that no two threads will get the same `bfreg` so for an unused `bfreg`, `reg` wouldn't be set only by the thread that gets it. Each dynamic `bfreg` is used only once and so the check would never evaluate to true. Just want to make sure I am not missing anything.

Reply:
It may be evaluate to true post mlx5_detach_dedicated_bf(), then it will be available back for further usage.

Reply:
Yes, makes sense. Thanks!